### PR TITLE
Update to v8.4.0

### PR DESCRIPTION
**Pull request by xylar**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by xylar** [comment]
@conda-forge-admin, please rerender

**Comment by xylar** [comment]
@zklaus, I guess this is really the version needed, since OSX ARM migration for ESMF happened for this version.  I'm not sure if we want to skip #58 and just go for this one.

And maybe ESMPy isn't of interest to you at all.

**Comment by xylar** [comment]
Also, this PR is just to update to 8.4.0, not to do anything with OSX ARM.  So just getting this to link correctly would be the first step, and likely doesn't need to wait for the new package build unless you think you've fixed things on the ESMF side in a way that would fix the linking problem here without any other changes needed.  Actually, maybe it's worth waiting a few hours to find out if that's the case before tinkering further...

**Comment by zklaus** [comment]
Agreed. But perhaps for now we can stick it out :smirk:. Btw, I am very much interested in ESMPy, cf ESMValGroup/ESMValTool#2568.

Let's try to get everything green here, but **not** merge, and see if we can apply any lessons learned to #58.

File: recipe/build.sh [hunk 1]
```diff
@@ -2,9 +2,20 @@
...
-${PYTHON} setup.py install --record record.txt
+
+${PYTHON} -m pip install .
+
+for shell in sh csh fish
+do
+  for act_deact in activate deactivate
+  do
+    act_deact_dir=${PREFIX}/etc/conda/${act_deact}.d
+    mdkir -p ${act_deact_dir}
```
[the remark this answers is not in the document]
Comment (xylar) [reply recipe/build.sh:18]: Oops, that's one I make a lot.

**Comment by xylar** [comment]
@zklaus, you should now be able to edit at will.

**Comment by zklaus** [comment]
@conda-forge-admin, please rerender

**Comment by zklaus** [comment]
@conda-forge-admin, please rerender

**Comment by zklaus** [comment]
There might be an actual error here? I attach the two relevant logfiles. They can be found in the artifacts as well, but they seem to be similar for all sub-builds, so here, only as an example, are the ones from the linux with openmpi:

- [esmpy8.4.0-petx1.test.txt](https://github.com/conda-forge/esmpy-feedstock/files/10187812/esmpy8.4.0-petx1.test.txt)
- [PET0.ESMF_LogFile.txt](https://github.com/conda-forge/esmpy-feedstock/files/10187813/PET0.ESMF_LogFile.txt)

The test that fails is `esmpy.test.test_api.test_regrid.TestRegrid testMethod=test_field_regrid_zeroregion_select_ndbounds`, where it first successfully regrids in memory, and then fails on writing weights to disk with
```
20221201 102756.604 ERROR            PET0 /home/conda/feedstock_root/build_artifacts/esmf_1669815663737/work/src/Infrastructure/Mesh/src/ESMCI_Mesh_Glue.C:3550 ESMCI_getlocalelemcoords() Arguments are incompatible  -  - elem_coordinates not found in Mesh
20221201 102756.604 ERROR            PET0 /home/conda/feedstock_root/build_artifacts/esmf_1669815663737/work/src/Infrastructure/Mesh/src/ESMCI_Mesh_Glue.C:3565 ESMCI_getlocalelemcoords() Arguments are incompatible  - Internal subroutine call returned Error
20221201 102756.604 ERROR            PET0 ESMC_Mesh.C:348 ESMC_MeshGetElemCoord() Arguments are incompatible  - Internal subroutine call returned Error
```
and follow-up error. Does anyone have any ideas?

**Comment by zklaus** [comment]
PS: I tried to find the upstream test runs for comparison, but couldn't. Does upstream routinely run the tests? Where would the logs be?

**Comment by xylar** [comment]
@rokuingh seems like the only one of us likely to be able to answer your questions, @zklaus.  Otherwise, it might be worth posting an issue upstream: https://github.com/esmf-org/esmf

**Comment by zklaus** [comment]
@conda-forge-admin, please rerender

**Comment by billsacks** [comment]
@rokuingh is on leave, so I'll try to answer regarding the upstream testing, with the caveat that I'm pretty new to ESMF and ESMPy myself.

ESMF's nightly test results can be found here:

http://earthsystemmodeling.org/esmf-test-summary/develop/index.html

We only run ESMPy testing on the cheyenne machine (with ESMF built with the intel compiler and with gfortran 10.1.0). You can hover over a table cell from the cheyenne-gfortran 10.1.0 testing to see a breakdown, showing that ESMPy tests are currently passing for us. You can click on the number in the table cell to see the full test artifacts.

The 8.4.0 release testing isn't shown in this table, but here is one set of results from the 8.4.0 testing:

https://github.com/esmf-org/esmf-test-artifacts-2022-10/tree/cheyenne/release_8.4.0/gfortran/10.1.0/g/mpt/2.23/esmpy

My understanding is that many of the apparent error messages in the log files are expected, I think because ESMPy does some failure / exception testing. The final line in https://github.com/esmf-org/esmf-test-artifacts-2022-10/blob/cheyenne/release_8.4.0/gfortran/10.1.0/g/mpt/2.23/esmpy/esmpy8.4.0-petx1.test doesn't indicate any failures, though (if I'm reading it correctly), so you could possibly compare your output with that one. If that doesn't help, you could email ESMF Support <esmf_support@ucar.edu> for help.

**Comment by zklaus** [comment]
Thanks, @billsacks, that is very helpful :+1:

**Comment by zklaus** [comment]
I should explain. It may look a bit desperate and irrational to just deactivate one test after another, but the underlying problem really is upstream: There are two main problems. One is that a significant number of tests (most of those that I am deactivating here) rely on helper methods that construct meshes or grids. These methods don't provide element coordinates, but the underlying C/C++/Fortran layer (yes, all of them are mashed together), demands those. Consequently, the underlying layer fails, but while this results in mere warning output on Linux, on Osx the compiler detects unhandled exceptions and barfs completely, which stops the Pytest tests in their tracks. As confirmed by looking at the test results linked to by billsacks, that's not a difference compared to upstream.

Mid- to long-term, it would be desirable to improve the situation upstream, but for now, we should probably just accept that a rather large number of tests is useless and deactivate them.

**Comment by xylar** [comment]
@zklaus, understood, thanks for the clarification.

**Comment by zklaus** [comment]
Please have a look. One thing that came up is that the `esmf.mk` file that comes in the **esmf** (not esmpy) package in `lib/esmf.mk` contains a bunch of not rewritten build env links. This seems to be messing with the nompi setup, which is why I had to patch that part the way I did. Is this something we should address in the esmf feedstock? Or is that file ok and we just need to deal with it here?

What do you think? Also, should we try to make this all work on the 8.3.0 branch before we merge here?

**Comment by xylar** [comment]
@zklaus, it seems worth trying to fix `esmf.mk` in the esmf feedstock.

I wouldn't wait to merge this, though.  I would say we merge this and work on esmf and also the 8.3.1 branch.  I can create an 8.3.x branch for that work to go to.

**Comment by xylar** [comment]
@zklaus, I'll merge as soon as you give me the go-ahead.

**Comment by zklaus** [comment]
@conda-forge-admin, please rerender

**Pull request merged**